### DISCUSSION TITLE
Icon sizes

### DISCUSSION THREAD

**Discussion by ghost** [sagb/alttab#163] [General]
Is there a way to choose larger icons? If I set -i 256x256 it does this strange bit where it will pull the wrong icons for somethings but those will scale correctly where as where it pulls the correct icons it seems to limit itself to icon sizes of 64x64 then scales them up to become blurry.

So for example I have some terms with custom window classes and icons to match. Alttab grabs those icons but after 64x64 it upscales and they are ugly and blurry despite them having versions from 16x16 to 512x512. Then KeypassXC it uses a stock icon despite my icon theme having icons for it the same as the terms but that icon scales cleanly.

Just for experimentation I've tried all the -s options to no change. I can't figure out why for some things it tops out with the 64x64 icons but then others where the win_class has a matching icon it ignores it and defaults to hicolor. The only thing I've noticed was xprop doesn't seem to see an icon with the hicolor choices.

**Comment by sagb** [comment]
Alttab always uses a single, fixed theme, either defined with the `-theme` option or defaults to 'hicolor'. It ignores all other themes. Does this clarify everything?

**Comment by ghost** [comment]
Well no because it doesn't touch on why it won't use larger icons than 64x64 and why it's ignoring some icons in the set theme.

Just to be clear this is my launch params/args
DISPLAY=:0.1 alttab -vp pointer -fg \#AAAAAA -w 1 -sc 1 -s 2 -theme m0dw3rks -bw 1 -bc \#464646 -frame \#FFB600 -i 100x64 -t 100x100 -inact \#121212 &
DISPLAY=:0.0 alttab -vp pointer -fg \#AAAAAA -w 1 -sc 1 -s 2 -theme m0dw3rks -bw 1 -bc \#464646 -frame \#FFB600 -i 100x64 -t 100x100 -inact \#121212 &

What I'd like is (-i 256x256 -t 256x300|-i 128x128 -t 128x140) but all the icons are blurry due to scaling rather than using the 256x256 | 128x128 versions of the icons.

The issue of the ignored icon is stranger as I replaced the hicolor icons with the ones from my theme still it pulled the "original" icon after restarting from somewhere else. Xprop says the window has no set icon so rather than using the WM_CLASS in the set theme like the man page talks about it seems to do something else. Given all the hicolors are scalable svg's kinda says why this icon is the only one that scales over 64x64 but my icon set also has scalable so again why does alttab give up around 64x64 and never look for the matching size or switch to the scalable?

**Comment by sagb** [comment]
Alttab uses only PNG and XPM icons, not SVG. Does this explain everything?
If not, please start it with the `-vv` option and observe which icons it uses:
```
alttab -vv 2>&1 | grep -i xpdf
```
where `xpdf` is WM_CLASS of the app in question.

**Comment by ghost** [comment]
In the icon theme I have set 16 to 512 are png, there are xpm for applications (because my WM is as old as I am heh) and svg under "scalable."

When running with -vv added it doesn't show/tell what icons are being used (or from where). Just an icon id and size along with the win title not the class.
```
0: 2400004 (lvl 0, icon 60817413 (64x64)): Win-Title
1: 2800004 (lvl 0, icon 60817414 (64x64)): Win-Title
2: 2c00004 (lvl 0, icon 60817415 (64x64)): Win-Title
3: 440000a (lvl 0, icon 60817416 (64x64)): Win-Title
4: 3800009 (lvl 0, icon 60817417 (64x64)): Win-Title
5: 460000f (lvl 0, icon 60817418 (128x128)): Win-Title
6: 5800014 (lvl 0, icon 60817419 (128x128)): Win-Title
7: 4a0003e (lvl 0, icon 60817420 (64x64)): Win-Title
8: 6800099 (lvl 0, icon 60817421 (64x64)): Win-Title
9: 6600099 (lvl 0, icon 60817422 (64x64)): Win-Title
```

I removed my xpm and tested and then rebuilt them and tested again with the same results. It grabs a few at the correct 128x128 but the rest it keeps limiting 64x64.

**Comment by sagb** [comment]
At the very start of `alttab -vv`, before pressing Alt-Tab, are there no lines in the output that begin with "app"?
```
app "gvncviewer" [/usr/share/icons/hicolor/256x256/apps/gvncviewer.png] (256x256)
app "vncviewer" [/usr/share/icons/hicolor/256x256/apps/vncviewer.png] (256x256)
app "gnuplot" [/usr/share/icons/hicolor/64x64/apps/gnuplot.png] (64x64)
app "gvim" [/usr/share/icons/hicolor/48x48/apps/gvim.png] (48x48)
app "openhantek" [/usr/share/icons/hicolor/48x48/apps/OpenHantek.png] (48x48)
```

**Comment by ghost** [comment]
There is but it's listing things I don't have open or don't use and all pointing to /usr/share/pixmaps/
```
app "dosbox" [/usr/share/pixmaps/dosbox.png] (1x1)
app "leafpad" [/usr/share/pixmaps/leafpad.png] (1x1)
app "galculator" [/usr/share/pixmaps/galculator.xpm] (1x1)
app "gksu" [/usr/share/pixmaps/gksu.png] (1x1)
app "gksu-icon" [/usr/share/pixmaps/gksu-icon.png] (1x1)
app "gksu-root-terminal" [/usr/share/pixmaps/gksu-root-terminal.png] (1x1)
app "puddletag" [/usr/share/pixmaps/puddletag.png] (1x1)
app "soundconverter" [/usr/share/pixmaps/soundconverter.png] (1x1)
app "steam" [/usr/share/pixmaps/steam.png] (1x1)
app "steam_tray_mono" [/usr/share/pixmaps/steam_tray_mono.png] (1x1)
app "terminator" [/usr/share/pixmaps/terminator.png] (1x1)
app "garudalinux-logo" [/usr/share/pixmaps/garudalinux-logo.png] (1x1)
app "openmw-cs" [/usr/share/pixmaps/openmw-cs.png] (1x1)
app "openmw" [/usr/share/pixmaps/openmw.png] (1x1)
app "fish" [/usr/share/pixmaps/fish.png] (1x1)
app "wihotspot" [/usr/share/pixmaps/wihotspot.png] (1x1)
app "mplayer" [/usr/share/pixmaps/mplayer.png] (1x1)
app "htop" [/usr/share/pixmaps/htop.png] (1x1)
app "nvim" [/usr/share/pixmaps/nvim.png] (1x1)
app "archlinux-logo" [/usr/share/pixmaps/archlinux-logo.png] (1x1)
app "filled-xterm" [/usr/share/pixmaps/filled-xterm_48x48.xpm] (48x48)
app "mini.xterm" [/usr/share/pixmaps/mini.xterm_48x48.xpm] (48x48)
app "xterm" [/usr/share/pixmaps/xterm_48x48.xpm] (48x48)
app "kalu" [/usr/share/pixmaps/kalu.png] (1x1)
app "nvidia-settings" [/usr/share/pixmaps/nvidia-settings.png] (1x1)
app "vscodium" [/usr/share/pixmaps/vscodium.png] (1x1)
```

Above that it has
```
icon dir: /usr/share/icons/m0dw3rks
icon dir: /usr/local/share/icons/m0dw3rks
icon dir: $HOME/.icons/m0dw3rks
icon dir: $HOME/.local/share/icons/m0dw3rks
icon dir: /usr/share/pixmaps
icon dir: $HOME/.local/share/pixmaps
icon dirs: 217, files: 15461, apps: 26
```

**Comment by sagb** [comment]
On your system, alttab uses only pixmaps + icons from X window properties.

Let's try to figure out why it doesn't use images from freedesktop directories, such as the m0dw3rks theme. Please show the full path to any of the icons in this theme that you expect to be loaded, along with the output of `file <this path>`.

**Comment by ghost** [comment]
Well it _IS_ using my set theme icons but the wrong sizes. So all the stuff listed above doesn't make sense. Like I said, it's listing stuff not used/not running then showing it's using pixmap fles for them but none of that stuff is relevant.

So if I'm on an XScreen/Workspace with Waterfox running and hit alttab it shows $HOME/.local/share/icons/m0dw3rks/apps/64/waterfox.png even if I have it set with -i to use /home/shed/.local/share/icons/m0dw3rks/apps/128/waterfox.png and thus it's blurry and terrible looking.

**Comment by sagb** [comment]
Alttab doesn't load icons from any of these files:
```
$HOME/.local/share/icons/m0dw3rks/apps/64/waterfox.png
/home/shed/.local/share/icons/m0dw3rks/apps/128/waterfox.png
```
because it only searches for icons in `<theme directory>/HxW/apps/`.
If it appears like it uses your icons, then some of your icons might have made
their way into pixmaps or X properties somehow, I don't know.
Forget about this, they are not used properly.

Instead, to quickly fix things:
```
for x in 64 128 256 512 ; do
    t=$HOME/.local/share/icons/m0dw3rks
    mkdir $t/${x}x${x}
    ln -s $t/apps/$x $t/${x}x${x}/apps
done
```

**Comment by ghost** [comment]
The links do the trick in so far as alttab see things differently with the -vv output but it's still hit and miss on quality. Some icons look right, others still look terrible, blurry and scaled up. Looking at the output it is pulling the 128x128 but if I open the file in feh/sxiv it looks good, in alttab, blurry mess same as before. I see some icons as well have a black background added where the alpha should be. Other icons alpha is rendered OK but some not.

Anyhow thanks for looking into this. I'll just call it as 64x64 is the best that can be done. I'll have to restructure the icon builds too.

**Comment by sagb** [comment]
Does `-s 3 -i 128x128` look blurry?

**Comment by ghost** [comment]
With -s 3 most icons don't show at all and some of what do show have their alpha channel turned black. The ones that showed looked "crisp." I tested all -s options again and they all do some combination of small icons upscaled (blurry), some icons correct (crisp), some icons alpha channels turned black, some icons don't display.

Comically if I "break" my icons directory structure with the Open Desktop Standard the black alpha goes away and it still pulls the same icon.

I tried changing the icon theme to something else and restarting alttab...it's still somehow using my normal icon theme. This feels more like ghosting hunting than debugging heh.

**Comment by sagb** [comment]
I would recommend continuing to try `-s 3` because it's easier to debug, and you probably want to use your specific icon theme, not random X properties.

Alttab is very straightforward; it doesn't involve any obscure processes. For instance, it uses ONLY the theme specified with the `-theme` option, not your "system theme" or anything else.

With `-s 3`, you can be sure that it uses ONLY those icons which are listed in the "app ..." lines at the start. Take a look at these lines; everything you need is there. Just figure out why it doesn't meet your expectations.

Ensure the PNG files are indeed PNG format.
Good luck.

**Comment by ghost** [comment]
Heh they are png, this is my icon set so I know how it's built/rendered. As for going -s 3 it's not really an option. Most icons don't show at all, those that do have the blacked out alpha, it's a mess, despite the "app" lines saying it's using the right files. At least with -s 2 the icons show, the alpha is correct and as long as I don't go larger than 64x64 they look as they should.

**Comment by sagb** [comment]
**From the icons displayed in the "app" lines**, could you attach to the chat one that is not shown and another one with wrong alpha?

Also, `xwininfo -root|grep Depth`

**Reply by ghost** [reply]
Depth is 24 but that doesn't matter because the alpha is shown correctly as long as 64x64 is set. The icons that lose their alpha are for example my Geany icons, but only if I go above 64x64 and the directory structure is inverted i.e. 128x128/apps/geany.png

**Comment by sagb** [comment]
Once again, if you encounter a situation where the path to a valid icon is displayed in the 'app ...' line, yet alttab incorrectly renders this icon, then please attach the icon here to this conversation by using the 'attach' button.

**Reply by ghost** [reply]
![geany](https://github.com/sagb/alttab/assets/38122105/781d46a4-d803-435f-af78-ddcd9240dad4)

I will test a few things in a bit to see if the alpha is ditched (and the app line matches) if I keep things 64x64 with the inverted directory structure.

Just did a quick test while waiting on a compile. This time Geany looks fine but my Ranger icons alpha is nuked.
app "ranger" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/ranger.png] (128x128)
![ranger](https://github.com/sagb/alttab/assets/38122105/2dd02fe2-8707-482c-9d83-7bdba453e9a6)

Tried again dropping back to 64x64, now Geany is back to losing alpha as is my leafpad icon.
app "leafpad" [$HOME/.local/share/icons/m0dw3rks/64x64/apps/leafpad.png] (64x64)
app "geany" [$HOME/.local/share/icons/m0dw3rks/64x64/apps/geany.png] (64x64)
![geany](https://github.com/sagb/alttab/assets/38122105/7188c599-ef78-427f-98c4-b470a0b767b6)
![leafpad](https://github.com/sagb/alttab/assets/38122105/b347725b-7e4e-444f-a89a-5903bc8a6b94)

So looks like as long as the inverted directory structure (ala open desktop standard) of size/category is in play it's a whack-a-mole game of which icons it will screw up...or not show at all (though all are showing right now).

I've been playing around with things and have another fun one.

So I launched 128x128 again and in the APP list I have
app "palemoon" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/palemoon.png] (128x128)
but when I trigger alttab that is NOT the icon it displays. It looks to be grabbing 64x64 from highcolor.

**Comment by sagb** [comment]
I just checked your 128x icons in alttab, and they are displayed nicely, including the alpha channel:

![r128](https://github.com/sagb/alttab/assets/4435811/536aa9ce-78ac-493d-8a2d-733d5f19c5e6)

I don't quite understand what do you mean by "inverted" directory structure; only one directory structure is recognized by alttab, and it's a standard, so I recommend to stick to it (though you may have reasons for another scheme, I don't know).

**Reply by ghost** [reply]
See my other stuff above where the APP line says it's using one icon but alttab displays another.
So I launched 128x128 again and in the APP list I have but when I trigger alttab that is NOT the icon it displays. It looks to be grabbing from highcolor.

app "palemoon" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/palemoon.png] (128x128)
(When I alttab)
7: 4800099 (lvl 0, icon 37749015 (48x48)): Workout System 17 - Pale Moon
8: 5000099 (lvl 0, icon 37749016 (48x48)): New Mail - Pale Moon

Also this has the correct size but it IS NOT using the icon listed in the app line, it's using the default TOR icon.
app "tor-browser" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/tor-browser.png] (128x128)
9: 600003e (lvl 0, icon 37749150 (128x128)): Polemount-singlephase-closeup.jpg (JPEG Image, 1024 × 1545 pixels) — Scaled (64%) — Tor Browser

Inverted because my icons are built with the structure of APPS/SIZE - thus "inverting" means SIZE/APPS.

I've tried to pipe alttab output to a log but it won't play nice. In addition to that I've tried to screen cap thing and it never captures.

Glad the icons work for you, sometimes they work here too, but that's the catch, alttab plays "whack-a-mole" with what it will trip up on. As I said several messages ago - just call this, it's like chasing ghosts rather than debugging and as long as I remove the inverted directory structure and stay 64x64 alttab behaves.

**Comment by ghost** [comment]
I will add one last thing. If you really want to figure this out note something I said way back about when you actually trigger alttab. Lines like "8: 5000099 (lvl 0, icon 37749016 (48x48)): New Mail - Pale Moon" The "icon" is denoted as an ID which is useless (at least to me). If you want to see the discrepancies between the APP lines and the actual triggers you should have alttab output the path and file it's *actually* loading to display. With the Tor Browser example the sizes line up so "on paper" it looks like it's pulling the right icon. You don't see the issue unless you visually see the wrong icon when you trigger. You could see in the logs it says one thing on the App line then grabs something totally different to show the user. That won't help (in theory) why it magically starts to butcher alpha channels for me when the directory structure is inverted but there is definitely some little gotcha happening here and lots of questions.

**Comment by sagb** [comment]
To clarify where the icons came from in these lines,

> 7: 4800099 (lvl 0, icon 37749015 (48x48)): Workout System 17 - Pale Moon
> 8: 5000099 (lvl 0, icon 37749016 (48x48)): New Mail - Pale Moon

I've added some debug in master branch. Use: `./configure CFLAGS="-DICON_DEBUG=1"` to build it.

**Reply by ghost** [reply]
[Exactly,](
https://github.com/sagb/alttab/assets/38122105/ea14783c-1ea2-4328-b0c9-a5981d5535e8
) right now alttab says one thing on the App line then does another and no way to tell how we went from A to B in the output. It's acting like a shady street vendor. He says it's a genuine Rolex but when you see its the face says Rollex. ;) Trust me bro, icon is legit!

Try to build in a few and report back.

With my normal directory structure
Found 3590 elements in _NET_WM_ICON (5000099, m0dMAIL v0.2 - Pale Moon)
using 48x48 _NET_WM_ICON icon for 5000099
depth=24

With the inverted (so I get the app lines)
app "palemoon" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/palemoon.png] (128x128)
app "palemoon" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/palemoon.png] (128x128)
6: 4800099 (lvl 0, icon 102760461 (48x48)): Workout System 17 - Pale Moon
   from _NET_WM_ICON
7: 5000099 (lvl 0, icon 102760462 (48x48)): m0dMAIL v0.2 - Pale Moon
   from _NET_WM_ICON

app "tor-browser" [$HOME/.local/share/icons/m0dw3rks/128x128/apps/tor-browser.png] (128x128)
9: 600003e (lvl 0, icon 102760464 (128x128)): breaker panel at DuckDuckGo — Tor Browser
   from _NET_WM_ICON

0: 460000a (lvl 0, icon 102760457 (128x128)): ranger
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/ranger.png <-- Alpha is blacked out

Since I can't get screen shots to capture I did a quick screen record. You can see both wrong icons from what the app lines say as well as various icons where the alpha channel is blacked out.

https://github.com/sagb/alttab/assets/38122105/ff25089c-d478-473e-84b2-951e59c27de8

**Comment by sagb** [comment]
Let's please refrain from inappropriate analogies. Alttab retrieves icons from various locations, which might appear unexpected to a user who hasn't understood the working algorithm and options.

From your output, it's evident that the icons come from X window properties, not from files. Now, launch alttab with the `-s 3` option. Where will the Pale Moon icon come from?

**Reply by ghost** [reply]
-s 3
https://github.com/sagb/alttab/assets/38122105/d0a3e24c-675b-49a6-a21b-86dc79ff191e

That is why I said -s 3 is a no go. You can also see two Sakura terms, one icon is shown, the other isn't. It's the same icon so what gives?!

**Comment by sagb** [comment]
What are shown in these lines?
> 9: 600003e (lvl 0, icon ...

**Reply by ghost** [reply]
```
0: e017f8 (lvl 0, icon 102760455 (128x128)): Unity - TheRefinery - The Refinery - PC, Mac & Linux Standalone - Unity 2020.3.29f1 Personal <OpenGL 4.5>
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/unity.png
1: 4e0002d (lvl 0, icon 0 (0x0)): Icon sizes · sagb/alttab · Discussion #163 — Waterfox
   
2: 1a00004 (lvl 0, icon 102760453 (128x128)): shed@pillar:~
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
3: 2600004 (lvl 0, icon 102760453 (128x128)): ~
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
4: 2200004 (lvl 0, icon 102760453 (128x128)): sakura
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
5: 1c00004 (lvl 0, icon 102760453 (128x128)): shed@pillar:~
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
6: 520000f (lvl 0, icon 102760454 (128x128)): index.php - $HOME/www/www-fsrv/m0dw3rks/kalikitten - Geany
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/geany.png
7: 4c0002d (lvl 0, icon 0 (0x0)): Tuesday March 19th 2024 — Waterfox

0: 3a00003 (lvl 0, icon 62914569 (128x128)):  breaker-panels [$HOME/Documents/Blender/breaker-panels.blend] - Blender 4.0
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/blender.png
1: 3200009 (lvl 0, icon 62914567 (128x128)): quad
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/terminator.png
2: 6600007 (lvl 0, icon 0 (0x0)): SimpleScreenRecorder
   
3: 1e00004 (lvl 0, icon 62914565 (128x128)): shed@pillar:~
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
4: 2000004 (lvl 0, icon 62914565 (128x128)): shed@pillar:~/Desktop
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
5: 460000a (lvl 0, icon 62914566 (128x128)): ranger
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/ranger.png
6: 4a00014 (lvl 0, icon 0 (0x0)): .bash_aliases ($HOME) - Notepadqq
   
7: 4800099 (lvl 0, icon 0 (0x0)): Workout System 17 - Pale Moon
   
8: 5000099 (lvl 0, icon 0 (0x0)): m0dMAIL v0.2 - Pale Moon
   
9: 3e0000f (lvl 0, icon 62914568 (128x128)): session_start.sh - $HOME/Scripts/pillar - Geany
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/geany.png
10: 600003e (lvl 0, icon 0 (0x0)): breaker panel at DuckDuckGo — Tor Browser
   
11: 4200004 (lvl 0, icon 0 (0x0)): sakura
   
12: 5400008 (lvl 0, icon 62914570 (128x128)): Fluorescent Tube Light Bulbs.svg - Inkscape
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/org.inkscape.Inkscape.png
13: 6400004 (lvl 0, icon 62914565 (128x128)): sakura
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/sakura.png
14: 2400004 (lvl 0, icon 0 (0x0)): sakura
   
15: 3800004 (lvl 0, icon 62914571 (128x128)): *(Untitled)
   $HOME/.local/share/icons/m0dw3rks/128x128/apps/leafpad.png
```

**Reply by sagb** [reply]
(for Pale Moon windows)

**Reply by ghost** [reply]
4800099 (lvl 0, icon 0 (0x0)): Workout System 17 - Pale Moon
   
8: 5000099 (lvl 0, icon 0 (0x0)): m0dMAIL v0.2 - Pale Moon

**Comment by sagb** [comment]
Okay, no icons found.
Now, please run `xprop|grep CLASS`, click on Pale Moon window, show the output.

**Reply by ghost** [reply]
WM_CLASS(STRING) = "Navigator", "workout"

WM_CLASS(STRING) = "Navigator", "Tor Browser"
10: 600003e (lvl 0, icon 0 (0x0)): breaker panel at DuckDuckGo — Tor Browser
***No Icon
   WM_CLASS(STRING) = "sakura", "termTTL"   
11: 4200004 (lvl 0, icon 0 (0x0)): sakura
***No Icon
(again one Sakura shows, the other doesn't and both have custom win classes)

**Comment by sagb** [comment]
This means that for Pale Moon, alttab searches only for navigator.png and workout.png, and explains why it doesn't find anything. In short, your Pale Moon windows are not "app palemoon" for alttab.
For fix, you can make a symlink navigator.png -> palemoon.png.

**Reply by ghost** [reply]
That I can do, but then what about the blacked out alpha channels? Or the fact icons are missing from applications other instances have their icons for? The "random" I'll load for one, but not the other?
Don't forget there is also Tor Browser with default Win class, no icon. At least with -s 2 it loads the stock icon heh.

**Comment by sagb** [comment]
For blacked out icons: perform the same analysis, to realize where are they come from. If they come from X Window properties, then make them come from the sources you control (i.e., from files), you already know how.
If the alpha is still blacked out, then attach the icon in this chat.

I suggest we tackle the problems one at a time.

**Reply by ghost** [reply]
All the icons are coming from the same place...if it's loading MY icons, it's MY icons which have their alpha. I've already attached problematic icons and my entire icon set is on github.

Well I suggested we just leave it be. ;p Thanks for looking into things but as I've said many times -s 2 and no higher than 64x64 and it behaves. No missing icons, no blacked alpha, no blur.
